resharper/resharper-unity/src/CSharp/Psi/CodeAnnotations/CustomCodeAnnotationProvider.cs

```diff
@@ -183,22 +174,123 @@ public ICollection<IAttributeInstance> GetSpecialAttributeInstances(IClrDeclared
+            };
         }
 
         private IPsiModule GetModule(IClrDeclaredElement element)
         {
             return myExternalAnnotationsModuleFactory.GetPsiModule(element.Module.TargetFrameworkId) ?? element.Module;
         }
+
+        // We can't always use SpecialAttributeInstance because it resolves the IClrTypeName against the default
+        // (project) context. As luck would have it, Unity ship *some* annotations, so it will work for some
+        // annotations. However, if we need to use any annotations that aren't shipped (e.g. ValueRangeAttribute) then
+        // it will fail to resolve. We need to resolve first against the project, and then fallback to the
+        // ExternalAnnotations PsiModule.
+        private class AnnotationAttributeInstance : IAttributeInstance
+        {
+            private readonly IClrTypeName myClrTypeName;
+            private readonly IPsiModule myExternalAnnotationsModule;
+            private readonly IProject myProject;
+            private readonly AttributeValue[] myCtorArguments;
+
+            public AnnotationAttributeInstance(IClrTypeName clrTypeName, IPsiModule externalAnnotationsModule,
+                IProject project, AttributeValue[] ctorArguments)
+            {
+                myClrTypeName = clrTypeName;
+                myExternalAnnotationsModule = externalAnnotationsModule;
+                myProject = project;
+                myCtorArguments = ctorArguments;
+            }
+
+            public IClrTypeName GetClrName() => myClrTypeName;
+            public string GetAttributeShortName() => myClrTypeName.ShortName;
+
+            public IDeclaredType GetAttributeType()
+            {
+                return TypeFactory.CreateTypeByCLRName(myClrTypeName, NullableAnnotation.NotNullable,
+                    myExternalAnnotationsModule);
+            }
+
+            public IConstructor Constructor
+            {
+                get
+                {
+                    var typeElement = GetAttributeType().GetTypeElement();
+                    if (typeElement == null)
+                    {
+                        using (CompilationContextCookie.OverrideOrCreate(
+                            myExternalAnnotationsModule.GetResolveContextEx(myProject)))
+                        {
+                            typeElement = GetAttributeType().GetTypeElement();
+                        }
+
+                        if (typeElement == null)
+                            return null;
+                    }
+
+                    foreach (var constructor in typeElement.Constructors)
+                    {
+                        if (myCtorArguments.Length == 0)
+                        {
+                            if (constructor.IsDefault) return constructor;
+
+                            continue;
+                        }
+
+                        var parameters = constructor.Parameters;
+                        if (parameters.Count == myCtorArguments.Length)
```

Comment:
that approach will not work for attributes with default parameters or if attribute arguments will be passed by name(order could be changed)

Reply:
I am not sure is it relevant to usage of your class or not

Reply:
This doesn't affect usage. The attribute instance I'm creating is simple, only intended to be used to create a small number of annotation attributes and uses an array of values to find a constructor, based on correct order and type. So it's not intended to be used with default or named parameters.